x-pack/plugins/infra/public/hooks/use_log_view.ts

```diff
@@ -5,118 +5,126 @@
+      ? state.context.logView
+      : undefined
   );
 
-  const [updateLogViewRequest, updateLogView] = useTrackedPromise(
-    {
-      cancelPreviousOn: 'resolution',
-      createPromise: logViews.putLogView.bind(logViews),
-      onResolve: setLogView,
-    },
-    [logViews]
+  const resolvedLogView = useSelector(logViewStateService, (state) =>
+    state.matches('checkingStatus') || state.matches('resolved')
+      ? state.context.resolvedLogView
+      : undefined
   );
 
-  const [loadLogViewStatusRequest, loadLogViewStatus] = useTrackedPromise(
-    {
-      cancelPreviousOn: 'resolution',
-      createPromise: logViews.getResolvedLogViewStatus.bind(logViews),
-      onResolve: setLogViewStatus,
-    },
-    [logViews]
+  const logViewStatus = useSelector(logViewStateService, (state) =>
+    state.matches('resolved') ? state.context.status : undefined
   );
 
-  const isLoadingLogView = loadLogViewRequest.state === 'pending';
-  const isResolvingLogView = resolveLogViewRequest.state === 'pending';
-  const isLoadingLogViewStatus = loadLogViewStatusRequest.state === 'pending';
-  const isUpdatingLogView = updateLogViewRequest.state === 'pending';
+  const isLoadingLogView = useSelector(logViewStateService, (state) => state.matches('loading'));
+  const isResolvingLogView = useSelector(logViewStateService, (state) =>
+    state.matches('resolving')
+  );
+  const isLoadingLogViewStatus = useSelector(logViewStateService, (state) =>
+    state.matches('checkingStatus')
+  );
+  const isUpdatingLogView = useSelector(logViewStateService, (state) => state.matches('updating'));
 
   const isLoading =
     isLoadingLogView || isResolvingLogView || isLoadingLogViewStatus || isUpdatingLogView;
 
-  const isUninitialized = loadLogViewRequest.state === 'uninitialized';
+  const isUninitialized = useSelector(logViewStateService, (state) =>
+    state.matches('uninitialized')
+  );
 
-  const hasFailedLoadingLogView = loadLogViewRequest.state === 'rejected';
-  const hasFailedResolvingLogView = resolveLogViewRequest.state === 'rejected';
-  const hasFailedLoadingLogViewStatus = loadLogViewStatusRequest.state === 'rejected';
+  const hasFailedLoadingLogView = useSelector(logViewStateService, (state) =>
+    state.matches('loadingFailed')
+  );
+  const hasFailedResolvingLogView = useSelector(logViewStateService, (state) =>
+    state.matches('resolutionFailed')
+  );
+  const hasFailedLoadingLogViewStatus = useSelector(logViewStateService, (state) =>
+    state.matches('checkingStatusFailed')
+  );
 
-  const latestLoadLogViewFailures = [
-    loadLogViewRequest,
-    resolveLogViewRequest,
-    loadLogViewStatusRequest,
-  ]
-    .filter(isRejectedPromiseState)
-    .map(({ value }) => (value instanceof Error ? value : new Error(`${value}`)));
+  const latestLoadLogViewFailures = useSelector(logViewStateService, (state) =>
+    state.matches('loadingFailed') ||
+    state.matches('resolutionFailed') ||
+    state.matches('checkingStatusFailed')
```

Comment:
**nit**: when matching against multiple states, the [state documentation](https://xstate.js.org/docs/guides/states.html#state-methods-and-properties) suggest using the `.some` iterator on a states array:
```js
['loadingFailed', 'resolutionFailed', 'checkingStatusFailed'].some(state.matches)
```
It may be useful to improve readability

Reply:
I chose that approach to improve type safety. If we were to rename a state, the compiler would call that out in the current implementation, but not in the `Array.some` variant. Adding type annotations to the array to ensure the compiler knows only state names are allowed here would probably be even less readable than unrolling the boolean disjunction.

In addition (in other cases, but not this specific selector) the `Array.some()` approach doesn't cause the compiler to narrow the typestate (and therefore context types) while the `state.matches()` disjunction does that too.

Overall I'd prefer to err on the side of more type safety here if that would be alright with you?

Reply:
It makes sense yep, this comment was based on the other thought of extracting the states into an enumeration, so in case of changing a state, the enum would have checked for it. Anyway, shaping the implementation in favour of type safety is totally worth it :)